developer_docs/semantic_stage.md

```diff
@@ -0,0 +1,184 @@
+The `Module` object must then be saved in the [`pyccel.ast.utilities.builtin_import_registery`](../pyccel/ast/utilities.py) dictionary.
+
+## Low-level Objects
+
+Python is a high-level language.
+As a result many implementation details are hidden to the user.
+In low-level languages it is necessary to know this information.
+In the semantic stage we must therefore identify these details and add objects representing them.
+
+The most obvious examples of this are array allocation and garbage collection.
+However, there are other examples so it is important to consider whether there is anything happening under the hood when adding a new `_visit_X` function.
+Often to handle these hidden details new variables must be created.
+In this case it is important to use the scope to avoid name collisions.
+These variables that are created should be tagged as `is_temp = True`.
+This allows Pyccel to differentiate between variables which appear in the code and should be preserved at all costs, and variables which are created by Pyccel and may be omitted if it leads to cleaner code.
+
+Additional objects can often appear in awkward places where they cannot be easily returned as a `CodeBlock`.
+This is the case for example if the object is needed to properly define something inside the right hand side of an `Assign`.
+As the right hand side of an Assign cannot be a `CodeBlock` the additional expressions must be collected outside the usual flow.
+The attribute `SemanticParser._additional_exprs` exists to hold these expressions.
+This object is a list of lists which is initialised and inserted in `_visit_CodeBlock`.
+A list of lists is necessary in case a `CodeBlock` can be found inside another (e.g. `a = [i for i in range]` where a `CodeBlock` contains the `Assign`, but another exists inside the `For` loop).
+
+In order to avoid problems arising from forgetfulness we try to add additional objects in the most general place possible.
+For example, allocation occurs in the function `SemanticParser._assign_lhs_variable`.
+Variable declarations are created in the printer when needed from the scope variables (this allows each language to place the decorators in the most appropriate location).
+
+## Object Tree
+
+All the objects in the Pyccel AST inherit from [`pyccel.ast.basic.Basic`](../pyccel/ast/basic.py).
+This super-class stores information about how the various objects are related.
+This allows the class to provide functions such as `get_user_nodes` (which returns all objects of a given type which use the node), `get_attribute_nodes` (which returns all objects of a given type which are used by the node), `is_attribute_of` (which indicates if the argument is used by the node), `is_user_of` (which indicates if the argument uses the node), and `substitute` (which allows all occurrences of an object in the node to be replaced by a different object).
```

Comment:
I think there is a mistake here:

- `self.is_attribute_of(arg)` indicates if the node `self` is used by the argument `arg`
- `self.is_user_of(arg)` indicates if the node `self` uses the argument `arg`

Your description says the opposite.

Reply:
Fixed 9b7f3758